### ISSUE TITLE
Comparison of a boolean expression with `== false` is wrongly "optimized"

### ISSUE THREAD

**Issue by vn971** [rust-lang/rust-clippy#4717]
Hi. I'm getting a clippy warning when I write code like this:
```
if something.conversion_function().another_conversion().validated() == false {...}
```

The current behavior is to propose re-writing the code like that:
```
if !something.conversion_function().another_conversion().validated() {...}
```

Let's compare the original and the proposed rewrite. I would like to show that the rewrite is actually worse than the original.

First and most important, the re-write forces you to keep more things in your mind. Assuming that you read code left-to-right, this is how you read it:
```
if !something.conversion_function().another_conversion().validated() {...}
```

* you have to not not miss the tiny symbol `!` that negates things, which would otherwise ruin your code in a hard-to-find fashion.
* keeping the negation in mind, consider what `something` is doing.
* keeping the negation in mind, consider what `conversion_function()` does to the previous result
* keeping the negation in mind, consider what `another_conversion()` does the the previous result
* keeping the negation in mind, consider what `validated()` does to the prev result
* now remember that we need to negate the final result, and negate it

In the original code, however, you would only ever required to keep one thing in mind at a time. You can read it left-to-right in the order of occurence of the underlying types and values. Humans are known to be weak in keeping multiple things at the same time, with a limit as low as 3-4 items: [[1]](https://www.livescience.com/2493-mind-limit-4.html) [[2]](http://theconversation.com/working-memory-how-you-keep-things-in-mind-over-the-short-term-75960)

From another perspective:
`!something.function().function().is_good()` <--- given 1 second to read the code, does it check that something is good or bad?

----

The explanation on the current warning is here: https://rust-lang.github.io/rust-clippy/master/index.html#bool_comparison

This example is given:
```
if x == true {} // could be `if x { }`
```
That is, of course, bloody obvious, and nobody proposes to write `== true`. The current warning might, however, be wrong in generalizing the situation to chained methods and `== false`.

**Comment by Manishearth** [comment]
I feel like this is a pathological case, and _usually_ that's a false positive, but in this case really it's just the suggestion that's not as good as it could be. The `if` block being confusing is a symptom of the condition being super long; the correct solution is to split it out into a well-named condition variable.

**Comment by vn971** [comment]
I don't understand the reason for closing the issue. Right now, clippy throws a warning and makes a suggestion that will lead to less readable code. To me, putting `!` in front is an anti-pattern in itself, because of how easy it is while reading it, if the text around ! is more than 20 characters.

Can you give some argument except "I don't like it"? While it might be true that some `if`-s are better to be re-written, clippy doesn't suggest it. It suggest re-writing into code that has worse readability, by reasons outlined above.

**Comment by vn971** [comment]
EDIT ^

**Comment by vn971** [comment]
* and EDIT another typo ^

**Comment by vn971** [comment]
> I feel like this is a pathological case, and usually that's a false positive

If it's usually a false positive, then the issue should not be closed, but instead considered a bug and `_ == false` expressions not being touched by clippy?

**Comment by Manishearth** [comment]
@vn971 You misunderstand, I'm saying that _usually_ pathological cases are false positives, but in this case I'm saying it isn't: Clippy has caught a problem in your code, it's suggestion isn't the best it could be though.

Reopening for making the suggestion change into a "please make this a variable" hint when the string is long.

**Comment by vn971** [comment]
@Manishearth can you give more insight into _why_ you think clippy caught a problem? I gave arguments above why I think it is the opposite. In short, it is that the prefixing `!` breaks left-to-right readability and requires more things to keep in memory. Put another way, syntactically, it wraps the logical flow of the program as `!(subprocess.status.success())` (the outer braces are not there in code, but they should be there mentally for a person that reads the code).

**Comment by Manishearth** [comment]
I'm saying that having an `if` statement that long is a problem in and of itself.

**Comment by vn971** [comment]
@Manishearth the warning is given not only in `if`-s though. `let` as well. And what is "that long" exactly? Clippy currently _always_ makes the suggestion that can either not change anything for really simple syntax trees, or make it worse for others. What's the point in having the lint if it's never good? Note that I'm only talking about `== false` here. Not `== true`, not `!= false`, not `!= true`. Those last usages are correctly warned about.

**Comment by Manishearth** [comment]
right, i'm saying that you shouldn't be having expressions that long, i was focusing on ifs because that was your example.

I disagree that the lint is never good, it's the kind of lint that often catches stuff after refactorings.

(I also disagree that `!` is that unreadable)

**Comment by vn971** [comment]
"the kind of lint often catches stuff after refactorings" is a statement I never disagreed with. Of course it can catch it, because it's focused on `== true`, `!= false` etc. For those uses it's perfect.

**Comment by vn971** [comment]
If you agree with the additional mental effort in keeping the negation for chained code, but still find it more readable, then I guess it would be nice for anybody else-s opinion before proceeding with the issue or closing it.

My concern is mostly correctness here, and I saw unnoticed-`!`-on-the-left-side mistakes being made more often than explicit-`== false`-on-the-right. Otherwise `!` would be shorter of course.

**Comment by Manishearth** [comment]
The reason I point out refactorings is to address your contention that fixing ==false is never good. I've had that come up quite often in refactorings.

**Comment by vn971** [comment]
> I have been talking about == false this whole time

Oh, so you remember that clippy caught refactoring mistake with `== false` exactly? I thought you meant for this lint in general. Sorry for repeating then. I personally probably wouldn't remember these details.

**Comment by Manishearth** [comment]
Yes (multiple times), but more in general, my point was that _such lints_ where it feels like "nobody would right code like that, the lint is useless" often find issues that are introduced when refactoring.

**Comment by vn971** [comment]
Note that I don't think this lint falls into the described category. On the contrary, it's very easy, as I believe, to fall under its scope. Only problem is that I think the "fix suggestion" makes it worse rather than better.

**Comment by llogiq** [comment]
@vn971 you argue from an emotional position. You think the code is worse, but is it? I'm quite sure we will find people who agree or disagree all right.

Manish has identified a possible reason why this might be the case and offered a mitigation that would inarguably improve the code and make the lint suggestion appropriate.

This may even be lintable.

So what are you arguing for? That we should make the lint `allow`ed because it may encounter cases like yours?

**Comment by vn971** [comment]
> So what are you arguing for? That we should make the lint allowed because it may encounter cases like yours?

Yes. Even more than that, I think `== false` is never a problem, even in short scenarios. Anything else like `!something == true` is, but not this one. It's one of the simplest expression for a "false" statement that is left-to-right readable and requires no additional context for remembering.

> and offered a mitigation that would inarguably improve the code and make the lint suggestion appropriate

But said suggestion is 100% offtopic, right? For example, the code in first message here is 82 characters long, and perfectly fits one line. It should probably be unaffected by such new linter, right?

> you argue from an emotional position

Something to think about for me then, if you think this way as well. I asked for reasoning on why the rewrite is thought to be good. I don't think this question was answered so far. But if two people think my judgement might be broken on emotional reasons, I'm drawing back and reverting the issue to be closed. If anybody else thinks that there is ground for lint changing here, feel free to re-open.

**Comment by solarane** [comment]
Personally, I much prefer !condition than condition == false. For one, it is much less typing. From the day I've started coding, I have yet to see a single style guide or tutorial that prefers condition == false over !condition, while the opposite seems to be pretty common. I've seen a lot of people complain (including myself) about condition == false and how it is unnecessarily verbose, but you are the first person I have seen complaining about !condition.

I'm not saying my (and many others', indubitably) position is the "correct" position. I'm just saying yours isn't either. And, when it comes to stylistic lints and preferences, it tends to fall to what the idiom is (i.e. what more people prefer). In this case, I think the positioj clippy currently takes is the idiom in many cases.

**Comment by flip1995** [comment]
I agree with what @solarane wrote. Writing `!cond` instead of `cond == false` is the preferred variant in most style guides and therefore also Clippy implemented it that way. Having `== true` or `== false` in your code was even taught as an anti-pattern in my programming introduction course back in the day. 

But not everyone likes the same code style, and preferring `== false` instead of `!` is totally fine (I mean it is more verbose, and if you like more verbose, than you should write it that way). Therefore you can just `#![allow]` this lint in your code, no harm done.

**Comment by vn971** [comment]
I understood that people are really very used to boolean prefixes. It's impossible to try overcome it.

For me it makes no sense and probably never will, because of the aforementioned hard-to-find bugs and broken left-to-right readability. But I shouldn't impose my own thinking on others, people-s habits are really important. Programming courses or style guides are mentioned above. At first they seemed irrelevant to me, because they don't answer the "WHY", but later I understood that people-s habits are reason enough. You don't have to argue why you're using some inconvenient date format, or time format, or natural language quirks.

In other words, I in last two messages, I still only saw "less typing" as an argument besides references to habits. This is again fine, I'm not feeling bad or demotivated about the decision. I just underestimated the habits.

**Comment by vn971** [comment]
EDIT ^

**Comment by vn971** [comment]
EDIT  ^

**Comment by Boiethios** [comment]
> Personally, I much prefer !condition than condition == false. For one, it is much less typing. From the day I've started coding, I have yet to see a single style guide or tutorial that prefers condition == false over !condition, while the opposite seems to be pretty common. I've seen a lot of people complain (including myself) about condition == false and how it is unnecessarily verbose, but you are the first person I have seen complaining about !condition.
> 
> I'm not saying my (and many others', indubitably) position is the "correct" position. I'm just saying yours isn't either. And, when it comes to stylistic lints and preferences, it tends to fall to what the idiom is (i.e. what more people prefer). In this case, I think the position clippy currently takes is the idiom in many cases.

FWIW, I hate the `!` operator with passion. It's just one character, and can be easily missed or mistaken with `l` or `I` when eyeball-parsing the code quickly. OTOH, the `== false` is obvious. When I was a tech lead, I asked my team not to use `!` and none complained. In my current job, basically everyone use `== false` and once again, none complains about it. It's definitely not an isolated thing.

I wonder why Rust has taken those weird sigils from C as logical operators while `not`, `and` , `or`, as used in python for example, are far more readable, but that ship has sailed.

To go back to this issue, I think that it conflates 2 different things in one lint: `== true` is definitely wrong, and I see no reason why someone would write that. `== false`, though, is acceptable, and I wish I could allow to write the latter, and not the former, with a clippy configuration. In fact, I would like a lint to forbid both `== true` and the use of the `!` operator.

---

For those who come here looking for a solution to this, mine is to import `core::ops::Not` and to type `.not()`.

**Comment by jonaskivi** [comment]
I have always thought that `!` is horrible (because it is easily missed) and `== false` is obvious and clear. I almost always prefer == false, and I've written a lot of code for many companies in my career. This kind of code suggestions should be removed, since they decrease code readability and are very opionated/style issues anyway. (`.not()` seems nice but should be in the language as a keyword. Let's hope it gets fixed in the future.)

`#[allow(clippy::bool_comparison)]` will remove the unnecessary warning, but it also remove `== true` warning, since those two are tied together.
